### PR TITLE
8329109: Threads::print_on() tries to print CPU time for terminated GC threads

### PR DESCRIPTION
**Pull request by reinrich**
*EDIT 2024-04-02:* Use `Threads::non_java_threads_do(ThreadClosure* tc)` to print non-java threads (commit b2dabc95d760496d4896b30ef1a44cc5c250189d). It uses thread list iterators that synchronize with thread creation and termination. If a non-java thread is about to terminate, it waits for in progress iterations (see [NonJavaThread::remove_from_the_list](https://github.com/openjdk/jdk/blob/47f33a59eaaffc74881fcc9e29d13ff9b2538c2a/src/hotspot/share/runtime/nonJavaThread.cpp#L97-L102)).

Testing:

The output of a minimal example with G1 is almost identical (attached below). The order of non-java threads can differ.

The fix passed our CI testing: JTReg tests: tier1-4 of hotspot and jdk. All of Langtools and jaxp. JCK, SPECjvm2008, SPECjbb2015, Renaissance Suite, and SAP specific tests.
Testing was done with fastdebug builds on the main platforms and also on Linux/PPC64le.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329109](https://bugs.openjdk.org/browse/JDK-8329109): Threads::print_on() tries to print CPU time for terminated GC threads (**Bug** - P3)

### Reviewers
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18518/head:pull/18518` \
`$ git checkout pull/18518`

Update a local copy of the PR: \
`$ git checkout pull/18518` \
`$ git pull https://git.openjdk.org/jdk.git pull/18518/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18518`

View PR using the GUI difftool: \
`$ git pr show -t 18518`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18518.diff">https://git.openjdk.org/jdk/pull/18518.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18518#issuecomment-2024590236)

### CODE REVIEW

**Comment by albertnetymk** [comment]
It's `Universe::heap()->gc_threads_do` that may contain terminated threads, not any other kinds of`NonJavaThreads`, right?

If so, can the fix be placed in `gc_threads_do`, where every collector ensures only active gc threads are visited.

(I see that David made a similar remark in the bug report.)

**Comment by albertnetymk** [comment]
> WatcherThread, NativeHeapTrimmerThread can also be terminated.

I see.

I notice that `Thread::print_on` contains `if (osthread() != nullptr) {` to skip some threads. Is it possible to expand this condition to skip zombie ones? (Or set `_osthread` to nullptr when a thread enters zombie state.)

> (Btw: G1CollectedHeap::stop() doesn't stop G1CollectedHeap::_cm. Likely an oversight)

`_cm` is not a thread; it's the data structure used by `_cm_thread`. Therefore, it shouldn't/can't be stopped.

**Comment by reinrich** [comment]
> > WatcherThread, NativeHeapTrimmerThread can also be terminated.
> 
> I see.
> 
> I notice that `Thread::print_on` contains `if (osthread() != nullptr) {` to skip some threads. Is it possible to expand this condition to skip zombie ones? (Or set `_osthread` to nullptr when a thread enters zombie state.)

A very simple alternative could be to check `thread->osthread()->get_state()` in [fast_cpu_time](https://github.com/openjdk/jdk/blame/7ac2f914aadf7fb12b7258e54efae959e15d6721/src/hotspot/os/linux/os_linux.cpp#L5170) and just return -1 if the thread isn't `INITIALIZED` yet or already `ZOMBIE`.

> > (Btw: G1CollectedHeap::stop() doesn't stop G1CollectedHeap::_cm. Likely an oversight)
> 
> `_cm` is not a thread; it's the data structure used by `_cm_thread`. Therefore, it shouldn't/can't be stopped.

It has `_concurrent_workers` but even though they seem to work concurrently to the mutator they are just `WorkerThread` and not `ConcurrentGCThread`. Only the latter type of thread can be stopped.

**Comment by reinrich** [comment]
Thread dump of minimal example with G1.
Baseline: [thread_dump_baseline.txt](https://github.com/openjdk/jdk/files/14837376/thread_dump_baseline.txt)
PR: [thread_dump_pr.txt](https://github.com/openjdk/jdk/files/14837377/thread_dump_pr.txt)

**Review by MBaesken — APPROVED** [review]
Looks like a good and simple fix, thanks !  The doc of NonJavaThread::Iterator says 'only live fully-initialized threads can be found in the list' and this is what we need here.

File: src/hotspot/share/runtime/threads.cpp [hunk 1]
```diff
@@ -1336,10 +1336,7 @@ void Threads::print_on(outputStream* st, bool print_stacks,
   }
 
   PrintOnClosure cl(st);
-  cl.do_thread(VMThread::vm_thread());
-  Universe::heap()->gc_threads_do(&cl);
-  cl.do_thread(WatcherThread::watcher_thread());
-  cl.do_thread(AsyncLogWriter::instance());
+  non_java_threads_do(&cl);
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/threads.cpp:1339]: This certainly looks neater and simpler and ensures all NJT's will be processed. There is additional synchronization involved in using the `NonJavaThreads::Iterator` that I had to look closely at but I think it is safe to use in this context.

**Review by dholmes-ora — APPROVED** [review]

**Review by albertnetymk — APPROVED** [review]

**Comment by reinrich** [comment]
Thanks for the reviews. I think this is ready now...
/integrate

**Pull request closed without merging**